torch_em/model/unetr.py

```diff
@@ -3,190 +3,60 @@
 import torch.nn.functional as F
 
 from typing import Tuple
-from functools import partial
-from torch_em.model.unet import Decoder, ConvBlock2d, Upsampler2d
 
-# we catch ImportErrors here because segment_anything and micro_sam should
-# only be optional dependencies for torch_em
-try:
-    from segment_anything.modeling import ImageEncoderViT
-    _sam_import_success = True
-except ImportError:
-    ImageEncoderViT = object
-    _sam_import_success = False
+from .unet import Decoder, ConvBlock2d, Upsampler2d
+from .vit import get_vision_transformer
 
 try:
     from micro_sam.util import get_sam_model
 except ImportError:
     get_sam_model = None
 
 
-class ViT_Sam(ImageEncoderViT):  # type: ignore
-    """Vision Transformer derived from the Segment Anything Codebase (https://arxiv.org/abs/2304.02643):
-    https://github.com/facebookresearch/segment-anything/blob/main/segment_anything/modeling/image_encoder.py
-    """
-    def __init__(
-        self,
-        in_chans: int = 3,
-        embed_dim: int = 768,
-        global_attn_indexes: Tuple[int, ...] = ...,
-        **kwargs
-    ) -> None:
-        if not _sam_import_success:
-            raise RuntimeError(
-                "The vision transformer backend can only be initialized if segment anything is installed."
-                "Please install segment anything from https://github.com/facebookresearch/segment-anything."
-                "and then rerun your code."
-            )
-
-        super().__init__(embed_dim=embed_dim, **kwargs)
-        self.chunks_for_projection = global_attn_indexes
-        self.in_chans = in_chans
-        self.embed_dim = embed_dim
-
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
-        x = self.patch_embed(x)
-        if self.pos_embed is not None:
-            x = x + self.pos_embed
-
-        list_from_encoder = []
-        for i, blk in enumerate(self.blocks):
-            x = blk(x)
-            if i in self.chunks_for_projection:
-                list_from_encoder.append(x)
-
-        x = x.permute(0, 3, 1, 2)
-        list_from_encoder = [e.permute(0, 3, 1, 2) for e in list_from_encoder]
-        return x, list_from_encoder[:3]  # type: ignore
-
-
-def window_partition(x: torch.Tensor, window_size: int) -> Tuple[torch.Tensor, Tuple[int, int]]:
```

Comment:
Just curious, why don't we need those any more?

Reply:
Ahha, it is just redundant code. These are parts of ViT from SAM (which I forgot to delete previously hehe)